src/backend/vulkan/src/xr.rs

```diff
@@ -0,0 +1,330 @@
+use ash::vk::{self, Handle};
+use hal::image;
+use std::{
+    ffi::{CStr, CString},
+    sync::{atomic::Ordering, Mutex},
+};
+
+#[cfg(feature = "use-openxr")]
+use once_cell::sync::Lazy;
+use std::sync::atomic::AtomicBool;
+
+use crate::{conv, native};
+
+#[cfg(feature = "use-openxr")]
+pub(crate) struct Instance {
+    instance: openxr::Instance,
+    system: openxr::SystemId,
+    required_device_extension_properties: Vec<CString>,
+    required_instance_extension_properties: Vec<CString>,
+    requirements: openxr::vulkan::Requirements,
+
+    session: Option<openxr::Session<openxr::Vulkan>>,
+    frame_waiter: Option<openxr::FrameWaiter>,
+    frame_stream: Option<openxr::FrameStream<openxr::Vulkan>>,
+    space: Option<openxr::Space>,
+    // FIXME add field for all -validations state, that'd track separate parts
+}
+
+#[cfg(feature = "use-openxr")]
+impl std::fmt::Debug for Instance {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "OpenXR[...]")
+    }
+}
+
+#[cfg(feature = "use-openxr")]
+pub(crate) static INSTANCE: Lazy<Mutex<Option<Instance>>> = Lazy::new(|| Mutex::new(None));
+static INSTANCE_SET: AtomicBool = AtomicBool::new(false);
+
+#[derive(Debug)]
+#[cfg(feature = "use-openxr")]
+pub struct OpenXR {}
+
+#[cfg(feature = "use-openxr")]
+#[derive(Debug)]
+pub enum Error {
+    Unknown,
+}
+
+#[cfg(feature = "use-openxr")]
+impl OpenXR {
+    pub fn configure(instance: openxr::Instance) -> Result<OpenXR, Error> {
```

Comment:
This is the definition of `openxr::Instance`:
```rust
pub struct Instance {
    inner: Arc<InstanceInner>,
}
```
its cloneable, and transferred from app to the gfx through wgpu. See: 
![screenshot](https://raw.githubusercontent.com/blaind/xrbevy/main/docs/screenshot1.png)

However, this brings a problem when dropping the instance. `openxr::Instance` has a `impl Drop` that goes through FFI to C code. 

Firstly, I wonder what happens on the app code, is the `Arc`-reference kept even though the instance is on GFX side? Where should it be dropped?

Same applies to all dependent openxr-structs that are instantiated. Although, currently all except session are `.take()`:n away in the `get_session_handles()` below. Session can also be taken away in current implementation.

Reply:
> I guess this also brings out the question about desired safety level of openxr in gfx- and what steps can be documented for end-users to ensure the safeness even if the api is unsafe?

gfx-hal is totally unsafe, no need to bother with trying to make things safe at this level. But at least it should have a clear way for users to write safe abstractions. I.e. like in Vulkan, you have to destroy device resources before destroying the device itself.